### ISSUE TITLE
Gradient of jax.lax.select is wrong with jnp.inf

### ISSUE THREAD

**Issue by ibab** [google/jax#2630]
The gradient of jnp.where (defined via `jax.lax.select`) seems to break when there is a  `jnp.inf` or `jnp.nan` in a non-active branch:
```python
def f(x):
  return jnp.where(x > 0, 0, x * jnp.inf)
jax.grad(f)(1.)
```
```
DeviceArray(nan, dtype=float32)
```
This causes code designed to guard against nans to not work, e.g.
```python
jnp.where(x <= 0, 0, jnp.sqrt(x))
```

I've taken a look at the gradient of `select` defined here: https://github.com/google/jax/blob/f3a57656edaaec4fb7ad56190c7f0ade4f351e93/jax/lax/lax.py#L2774
But I couldn't immediately see where the nan is coming from.

**Comment by ibab** [comment]
Okay after thinking about it a bit more, this should come from the chain rule containing `0 * inf`. Seems like a good reason to define `0 * inf == 0`, at least inside of the autograd machinery!

**Comment by hawkinsp** [comment]
I'd read the chain of stuff at: https://github.com/google/jax/pull/2447
and
https://github.com/tensorflow/probability/blob/master/discussion/where-nan.pdf

**Comment by mattjj** [comment]
Indeed one more bit of context: `lax._safe_mul` was a multiply where `0 * inf == 0`, as you can see from the `_safe_mul_translation_rule` that was deleted in #2447. #1052 has a lot of detail (linked from #2447).

So I think this is working as intended, though if there are ways to improve coverage of programs without introducing any silently-wrong results for other programs, I'd love to think through them with you!

**Comment by ibab** [comment]
Okay I've read through `__safe_mul`'s history, and indeed it seems to be a tricky problem.
It seems that we have two different types of zeros that we want to encode in gradient rules, one is "zero means zero" for which `0 * inf = 0` and one is "happens to be zero" for which `0 * inf = nan`.
This can be implemented using a separate boolean mask like you've explained in the linked issues.
The two types of zeros can come up at the same time like in this example:
```
def f(x):
  return jnp.sum(jnp.where([True, False], x, jnp.sqrt(x)**2)
jax.grad(jnp.array([0., 0.]))
```
I would like this to return `[1., nan]`, but the way JAX is implemented it can't distinguish between the two different purposes of zeros.
It would have been cool if the floating point standard had included a "safe zero".
Closing this since I can't come up with any suggestions for making the situation better :(
